ceph/ceph_admin/osd.py

```diff
@@ -0,0 +1,103 @@
+"""
+Module to deploy osd service and daemon(s).
+
+this module deploy OSD service and daemon(s) along with
+handling other prerequisites needed for deployment.
+
+"""
+import json
+import logging
+from time import sleep
+
+from ceph.ceph_admin.apply import Apply
+from ceph.ceph_admin.daemon_mixin import DaemonMixin
+
+logger = logging.getLogger(__name__)
+
+
+class OSD(Apply, DaemonMixin):
+    def __get_available_devices(self):
+        """
+        Fetch all available devices using "orch device ls" command
+
+        Returns:
+            device_list: list of nodes with available devices
+
+        device_list:
+            node1: ["/dev/sda", "/dev/sdb"]
+            node2: ["/dev/sda"]
+        """
+        out, _ = self.shell(
+            remote=self.installer,
+            args=["ceph", "orch", "device", "ls", "-f", "json"],
+        )
+
+        node_device_list = dict()
+        for node in json.loads(out):
+            if not node.get("devices"):
+                continue
+            devices = []
+            for device in node.get("devices"):
+                if device["available"] is True:
+                    devices.append(device["path"])
+
+            if devices:
+                node_device_list.update({node["addr"]: devices})
+
+        return node_device_list
+
+    def apply_osd(self, **config):
+        """
+        Deploy all available OSDs in the ceph cluster using
+        "ceph orch apply osd --all-available-devices"
+
+        Args:
+            config: config
+
+        config:
+            all_available_devices: boolean
+        """
+        node_device_list = self.__get_available_devices()
+
+        assert node_device_list, "No devices available to create OSD(s)"
+        config.update({"role_args": ["--all-available-devices"]})
+        super().apply("osd", **config)
+
+        # validate of osd(s)
+        interval = 5
+        timeout = self.TIMEOUT
+        checks = timeout / interval
+
+        while checks:
+            checks -= 1
+
+            out, _ = self.shell(
+                remote=self.installer,
+                args=["ceph", "orch", "ps", "-f", "json-pretty"],
+            )
+
+            out = json.loads(out)
+            daemons = [i for i in out if i.get("daemon_type") == "osd"]
+
+            deployed = 0
+            for node, devices in node_device_list.items():
+                count = 0
+                for dmn in daemons:
+                    if dmn["hostname"] == node and dmn["status_desc"] == "running":
+                        count += 1
+
+                logger.info(
+                    "%s %s/%s osd daemon(s) up... Retries: %s"
+                    % (node, count, len(devices), checks)
+                )
+                if count == len(devices):
+                    deployed += 1
+
+            if deployed == len(node_device_list):
+                return True
+            sleep(interval)
+
+        assert False, "OSDs are not up and running in hosts"
```

Comment:
`assert` statement is generally for testing. IMO, we should use custom exceptions via `raise`.